pkg/kubelet/pleg/generic.go

```diff
@@ -122,34 +165,66 @@ func (g *GenericPLEG) relist() {
 		return
 	}
 
+	// Build a new map.
+	podMap := buildPodMap(pods)
+
 	events := []*PodLifecycleEvent{}
-	containers := make(map[string]containerInfo, len(g.containers))
-	// Create a new containers map, compares container statuses, and generates
-	// correspoinding events.
-	for _, p := range pods {
-		for _, c := range p.Containers {
-			cid := c.ID.ID
-			// Get the of existing container info. Defaults to state unknown.
-			oldState := kubecontainer.ContainerStateUnknown
-			if info, ok := g.containers[cid]; ok {
-				oldState = info.state
-			}
+	// Process all currently visible pods.
+	for pid, cMap := range podMap {
+		oldCMap, ok := g.podMap[pid]
```

Comment:
Good point. I think it's better to initialize this explicitly to avoid confusion. Done.